### PR TITLE
KASAN for QEMUv8

### PR DESCRIPTION
**Pull request by jforissier**
Adding support for `CFG_CORE_SANITIZE_KADDRESS=y` on QEMUv8.
This is work in progress, because `make CFG_CORE_SANITIZE_KADDRESS=y CFG_CORE_ASLR=n` works but there is an exception on boot when `CFG_CORE_ASLR=y`.
It seems the issue is in the prologue of function `virt_to_phys()`.
```
(gdb) bt
#0  virt_to_phys (va=va@entry=0xbe240000) at core/mm/core_mmu.c:2269
#1  0x00000000c24292a4 in check_va_matches_pa (pa=151257088, pa@entry=3190030336, va=va@entry=0xbe240000) at core/mm/core_mmu.c:2286
#2  0x00000000c2429650 in check_va_matches_pa (va=0xbe240000, pa=3190030336) at core/mm/core_mmu.c:2283
#3  phys_to_virt_io (pa=3190030336, len=len@entry=4096) at core/mm/core_mmu.c:2385
#4  0x00000000c2429750 in io_pa_or_va (p=0xbe240000, len=4096) at core/mm/core_mmu.c:2430
#5  0x00000000c2410274 in pl011_init (pd=0xc24f46e0, pbase=151257088, uart_clk=1, baud_rate=115200) at core/drivers/pl011.c:146
#6  0x00000000c24072d8 in console_init () at core/arch/arm/plat-vexpress/main.c:72
#7  0x00000000c23fa130 in _start () at core/arch/arm/kernel/entry_a64.S:368
```

```
000000000e12e734 <virt_to_phys>:
virt_to_phys():
/home/jerome/work/optee_repo_qemu_v8/optee_os/core/mm/core_mmu.c:2269
 e12e734:       a9ae7bfd        stp     x29, x30, [sp, #-288]!
 e12e738:       d2915662        mov     x2, #0x8ab3                     // #35507
 e12e73c:       f2a836a2        movk    x2, #0x41b5, lsl #16
 e12e740:       910483e1        add     x1, sp, #0x120
 e12e744:       910003fd        mov     x29, sp
 e12e748:       a90153f3        stp     x19, x20, [sp, #16]
 e12e74c:       927be833        and     x19, x1, #0xffffffffffffffe0
 e12e750:       d1030261        sub     x1, x19, #0xc0
 e12e754:       a9025bf5        stp     x21, x22, [sp, #32]
 e12e758:       aa0003f4        mov     x20, x0
 e12e75c:       d343fc36        lsr     x22, x1, #3
 e12e760:       a90363f7        stp     x23, x24, [sp, #48]
 e12e764:       f8140262        stur    x2, [x19, #-192]
 e12e768:       900004c2        adrp    x2, e1c6000 <__func__.8+0x1140>
 e12e76c:       91348042        add     x2, x2, #0xd20
 e12e770:       f8148262        stur    x2, [x19, #-184]
 e12e774:       90000002        adrp    x2, e12e000 <core_mmu_unmap_pages+0x1e8>
 e12e778:       911cd042        add     x2, x2, #0x734
 e12e77c:       f8150262        stur    x2, [x19, #-176]
 e12e780:       d2880002        mov     x2, #0x4000                     // #16384
 e12e784:       f2a18d42        movk    x2, #0xc6a, lsl #16
 e12e788:       8b410c41        add     x1, x2, x1, lsr #3
 e12e78c:       3204d3e2        mov     w2, #0xf1f1f1f1                 // #-235802127
 e12e790:       b931c022        str     w2, [x1, #12736]  <===== EXCEPTION HERE, JUMPS TO SynchronousExceptionSP0
```
```
Thread 1 received signal SIGINT, Interrupt.
reset_vect_table () at core/arch/arm/kernel/entry_a64.S:689
689             b       SynchronousExceptionSP0
```

### CODE REVIEW

File: core/arch/arm/kernel/boot.c [hunk 1]
```diff
@@ -299,8 +299,10 @@ static void init_asan(void)
 	asan_tag_access(__pageable_start, __pageable_end);
 #endif /*CFG_WITH_PAGER*/
 	asan_tag_access(__nozi_start, __nozi_end);
+#if defined(__arm__)
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/boot.c:302]: `s/__arm__/ARM32/`

Comment (jforissier) [reply core/arch/arm/kernel/boot.c]: Done

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by jforissier** [comment]
> there is an exception on boot when `CFG_CORE_ASLR=y`.

Happens with QEMUv7 too

**Comment by jenswi-linaro** [comment]
OK, that rings a bell. Thanks for the links.

**Comment by jforissier** [comment]
Tags applied, new commit: "core: asan: KASAN and ASLR are incompatible".

**Comment by jenswi-linaro** [comment]
Regarding "core: asan: KASAN and ASLR are incompatible". It might be possible to support KASAN with ASLR if we can keep the shadow area at the expected offset, but it would at least be much more flexible with a dynamic approach.

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Pull request merged**